### DISCUSSION TITLE
Unable to wire RDF4JTemplate into my linked data app

### DISCUSSION THREAD

**Discussion by mudduvenkatesh** [eclipse-rdf4j/rdf4j#4509] [General]
```
Application.yml - configured
rdf4j:
  spring:
    repository:
      remote:
        manager-url: http://localhost:7200
        name: CustomerMasterDataRepo
@Configuration
@Import(RDF4JConfig.class)
//@ComponentScan(
//        value = "io.ontollogix.apps.linkedmdm.cmda.dao", includeFilters = @ComponentScan.Filter(type = FilterType.ASSIGNABLE_TYPE, classes = RDF4JDao.class))
public class CmdaAppConfig {
    @Bean
    public DashboardDao getDashboardDao(@Autowired RDF4JTemplate template) {
        return new DashboardDao(template);

    }
}
```

Starting spring boot app causes the dependency error:

```
Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'getRdf4JTemplate' defined in class path resource [org/eclipse/rdf4j/spring/RDF4JConfig.class]: Unsatisfied dependency expressed through method 'getRdf4JTemplate' parameter 0: Error creating bean with name 'getRepositoryConnectionFactory' defined in class path resource [org/eclipse/rdf4j/spring/RDF4JConfig.class]: Unsatisfied dependency expressed through method 'getRepositoryConnectionFactory' parameter 0: No qualifying bean of type 'org.eclipse.rdf4j.repository.Repository' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}
2023-04-16T21:20:48.857-04:00  INFO 85998 --- [           main] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2023-04-16T21:20:48.887-04:00  INFO 85998 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger :
```

**Comment by abrokenjester** [comment]
I'm not immediately sure what could be wrong here, but at first glance it looks like it can't find your bean definition. Are you quite your config file is in the right place and being picked up by spring?

**Comment by nick-manasys** [comment]
I got a similar error when I tried to upgrade my project to use spring-boot-starter-parent 3.1.2 (from 2.7.12). The project compiled but I got a similar error when I ran it.

**Comment by hmottestad** [comment]
The spring components in RDF4J 4.x.x use spring-boot version 2.7.7. I don't believe that the spring developers have made 3.x.x compatible with 2.x.x.

We would appreciate any help you can offer in migrating our spring components to spring-boot 3.1.x.

**Reply by dennisvang** [reply]
@hmottestad Do you know if this is still relevant for rdf4j-spring 5.1.0 and spring-boot-starter-parent 3.4.1?

I'm getting a similar `No qualifying bean of type 'org.eclipse.rdf4j.repository.Repository' available` error when using a `SimpleRDF4JCRUDDao`, with `rdf4j.spring.repository.inmemory.enabled=true` in `application.properties`, following the [docs]. 

The error seems to go away when I define a bean that returns a new `SailRepository`, but I'm not sure if that's the right way to do it.

[docs]: https://rdf4j.org/documentation/programming/spring/#simplerdf4jcruddao

**Reply by hmottestad** [reply]
I haven't worked enough with the spring-boot integration to know if what you're doing works or not. I guess you could write some tests and see if all your use cases work as expected. Would also throw in some tests for failure scenarios to see if those are handled correctly. 

We are unfortunately stuck on the older spring-boot version. The newer version forces us to upgrade a lot of dependencies. Biggest job here is to migrate all the elasticsearch code to the latests client, and then to migrate all the tests to test containers (or whatever is the recommended approach for testing against elasticsearch). I'm still fairly grouchy at elastic for switching client frameworks so often and not keeping any of the old ones working. Also quite annoyed at how little they care for sticking to semantic versioning. Twice they've released new patch releases with breaking changes without any warnings or deprecations.

**Reply by dennisvang** [reply]
>I haven't worked enough with the spring-boot integration to know if what you're doing works or not. I guess you could write some tests and see if all your use cases work as expected. Would also throw in some tests for failure scenarios to see if those are handled correctly.

@hmottestad Thanks for the quick reply. I'll add some more tests and see how far I can get.

**Reply by dennisvang** [reply]
Although defining a repository bean does allow the app to start without errors, I now run into a "no qualifying bean of type ... TransactionManager" when trying to use the repository. Explicitly setting `@EnableTransactionManagement` and/or configuring `rdf4j.spring.tx.enabled=true` makes no difference.

**Reply by fkleedorfer** [reply]
Could you post the whole stack trace please? And maybe a link to the code if it is a public project?

**Comment by fkleedorfer** [comment]
I hope to be able to upgrade rdf4j-spring soonish. It's on my list. I assume I get to it  before end of february, maybe earlier.
